swarm/src/lib.rs

```diff
@@ -372,7 +372,16 @@ where
     /// Depending on the underlying transport, one listener may have multiple listening addresses.
     pub fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<io::Error>> {
         let id = ListenerId::next();
-        self.transport.listen_on(id, addr)?;
+
+        if let Err(e) = self.transport.listen_on(id, addr) {
+            self.behaviour
+                .on_swarm_event(FromSwarm::ListenerError(behaviour::ListenerError {
+                    listener_id: id,
+                    err: &e,
+                }));
+            return Err(e);
+        }
+
         self.behaviour
             .on_swarm_event(FromSwarm::NewListener(behaviour::NewListener {
```

Comment:
This should happen as well for the `ToSwarm` case I guess?

Kind of the most important bit :D

This makes me think that:

1. We should extract this into a `add_listener` function.
2. We should add a regression test for this functionality: a `NetworkBehaviour` that creates a listener, we connect to it and then remove it again. Prob best to use the `MemoryTransport` for it.

Reply:
Yes it should. I was wondering about last night prior to pushing up the code. Originally, I was calling `Swarm::listen_on`, which would have emitted the event (which worked in prior test), but since I was calling it directly from the transport, it would never emit the event (didnt get the chance to test). Glad that was caught now than later. 

As for your thoughts:
1. This is for calling `NetworkBehaviour::on_swarm_event`, correct? Also, If we are to add a `add_listener` function, would this be apart of `Swarm` and accessible by the end user or a private function? I assume the latter.
2. Should this test be located in the current crate?

Reply:
>     1. This is for calling `NetworkBehaviour::on_swarm_event`, correct? Also, If we are to add a `add_listener` function, would this be apart of `Swarm` and accessible by the end user or a private function? I assume the latter.

Yes a private function. We already have `remove_listener` so just a counter-part to that.

>     2. Should this test be located in the current crate?

I'd suggest we make it a test in `swarm/tests`. Where possible, I think our tests should use the public API and we can enforce that by putting them into the `tests` directory.

Reply:
>     1. This is for calling `NetworkBehaviour::on_swarm_event`, correct? Also, If we are to add a `add_listener` function, would this be apart of `Swarm` and accessible by the end user or a private function? I assume the latter.

Yes a private function. We already have `remove_listener` so just a counter-part to that.

>     2. Should this test be located in the current crate?

I'd suggest we make it a test in `swarm/tests`. Where possible, I think our tests should use the public API and we can enforce that by putting them into the `tests` directory.